service/mapper/transitions.go

```diff
@@ -0,0 +1,337 @@
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/optakt/flow-dps/models/index"
+)
+
+type TransitionFunc func(*State) error
+
+func Bootstrap(root Root, load Loader) TransitionFunc {
+	return func(s *State) error {
+
+		// Bootstrapping should only happen when the state is empty.
+		if !Empty(s) {
+			return fmt.Errorf("invalid state for bootstrap")
+		}
+
+		// We always need at least one step in our forest, which is used as the
+		// stopping point when indexing the payloads since the last finalized
+		// block. We thus introduce an empty tree, with no paths and an
+		// irrelevant previous commit.
+		empty := trie.NewEmptyMTrie()
+		ok := s.forest.Save(flow.DummyStateCommitment, empty, nil)
+		if !ok {
+			return fmt.Errorf("could not save empty tree")
+		}
+		prev := flow.StateCommitment(empty.RootHash())
+		s.last = prev
+
+		// Then, we can load the root height and apply it to the state. That
+		// will allow us to load the root blockchain data in the next step.
+		height, err := root.Height()
+		if err != nil {
+			return fmt.Errorf("could not get root height: %w", err)
+		}
+		s.height = height
+
+		// Next, we will load our checkpoint tree and add it as the step
+		// following the first empty tree. This will ensure that we index all
+		// paths within the root tree.
+		checkpoint, err := load.Checkpoint()
+		if err != nil {
+			return fmt.Errorf("could not read checkpoint: %w", err)
+		}
+
+		// There is no shortcut to get all the paths currently within the tree,
+		// so in order to get all paths for the checkpoint tree, we do a depth
+		// first iterative traversal.
+		paths := make([]ledger.Path, 0, len(checkpoint.AllPayloads()))
+		queue := deque.New()
+		root := checkpoint.RootNode()
+		if root != nil {
+			queue.PushBack(root)
+		}
+		for queue.Len() > 0 {
+			node := queue.PopBack().(*node.Node)
+			if node.IsLeaf() {
+				path := node.Path()
+				paths = append(paths, *path)
+				continue
+			}
+			if node.LeftChild() != nil {
+				queue.PushBack(node.LeftChild())
+			}
+			if node.RightChild() != nil {
+				queue.PushBack(node.RightChild())
+			}
+		}
+
+		// We need to sort this, otherwise the retrieval from the tree later
+		// when we index (with unsafe read) might fail to work properly.
+		sort.Slice(paths, func(i int, j int) bool {
+			return bytes.Compare(paths[i][:], paths[j][:]) < 0
+		})
+
+		// Now that we have the path, we are ready to add the checkpoint step.
+		ok = s.forest.Save(prev, checkpoint, paths)
+		if !ok {
+			return fmt.Errorf("could not save checkpoint tree")
+		}
+
+		// We have now successfully bootstrapped. However, no blockchain data
+		// has been indexed, so we pretend like we just indexed the empty tree
+		// and the forward transition will load our block data.
+		s.state = stateIndexed
+
+		return nil
+	}
+}
+
+func Update(feed Feeder, options ...func(*Config)) TransitionFunc {
+	cfg := DefaultConfig
+	for _, option := range options {
+		option(&cfg)
+	}
```

Comment:
Why pass the options to the transition func builders rather than have them in the state? Is it to make sure that the state itself only contains state and not configuration? I don't have anything against it, I'm just wondering.